design/one-pager-consuming-k8s-clusters.md

```diff
@@ -0,0 +1,270 @@
+is appropriate for Kubernetes clusters to be modeled in the same manner as other
+managed services due to the fact that they are presented by the provider in the
+same fashion. However, in Crossplane, Kubernetes clusters are unique to other
+managed service offerings in that they can also be used as a target for running
+workloads.
+
+Workloads are provisioned using the `KubernetesApplication` type. It is
+namespace-scoped and is scheduled to a target cluster using labels to match a
+`KubernetesCluster` claim in its namespace. Because managed resources may only
+be bound to a single claim, the claim is namespaced, and `KubernetesApplication`
+resources can only be scheduled to `KubernetesCluster` claims in their
+namespace, any Kubernetes managed resource offering may only be consumed (i.e.
+deployed to) from a single namespace.
+
+In order to fully conceptualize the usage of Kubernetes clusters in Crossplane,
+it is helpful to separate the notions of a **Kubernetes cluster as a managed
+service** and a **Kubernetes cluster as a target for deploying workloads**. With
+this distinction made, we can assert that this proposal in no way modifies the
+managed service notion of a Kubernetes cluster, but instead formalizes this
+concept of a Kubernetes cluster as a deployment target.
+
+## Goals
+
+Because it is unlikely that Crossplane users will always want a 1-to-1
+namespace-to-cluster consumption model, it is desirable to support deploying
+workloads in any namespace to any Kubernetes cluster. However, it is also
+desirable to continue to treat managed Kubernetes services as a managed resource
+that is supplied by cloud providers. For this reason, any solution should be
+additive to the existing model.
+
+As a secondary goal for this design, it should be possible to use existing
+Kubernetes clusters (wherever they may be running) that were not provisioned by
+Crossplane as targets for deploying workloads. This concept has frequently been
+referred to as a "Bring Your Own Cluster" scenario.
+
+## Proposal
+
+This proposal can be separated into two main categories:
+1. The implementation of a new `WorkloadTarget` custom resource.
+1. The modification of the scheduling behavior for `KubernetesApplication`.
+
+### WorkloadTarget
+
+The purpose of the `WorkloadTarget` resource is to effectively "publish"
+Kubernetes clusters for usage in a namespace. It is namespace-scoped and can
+only reference Kubernetes cluster managed resources or the Kubernetes `Provider`
+type, both of which are cluster-scoped. An example of a `WorkloadTarget` in
+namespace `my-app-team` that is referencing a GKE cluster managed resource could
+look as follows:
+
+```yaml
+apiVersion: workload.crossplane.io/v1alpha1
+kind: WorkloadTarget
+metadata:
+  namespace: my-app-team
+  name: enable-dev-k8s
+  labels:
+    env: dev
+clusterRef:
+  kind: GKECluster
+  apiVersion: compute.gcp.crossplane.io/v1alpha3
+  name: dev-k8s-cluster
```

Comment:
Hmm, I am not sure that this would ever happen. Names of Kubernetes objects are immutable so the cluster object would have to be deleted and then recreated, in which case I think a new `WorkloadTarget` (naming pending update haha) should have to be published again.

Reply:
But we can't guarantee that the old `WorkloadTarget` will be deleted, too. Someone can delete `Provider`/`KubernetesCluster` and after a day or two, create another one with the same name. Old `WorkloadTarget` will suddenly be working again with a different cluster whose intent may or may not be to be a target for that specific `WorkloadTarget`.